### PR TITLE
Adopt sphinx-apidoc and retire complex sphinxext

### PR DESCRIPTION
**Pull request by tkknight**
### Description
Retire the old and convoluted sphinx extensions to generate the Iris API documentation.

- Uses [sphinx-apidoc](https://github.com/sphinx-contrib/apidoc) instead
- The indented blank line is now gone that appeared at the top of some API doc pages
- The "**In this module:**" section  is now not present (part of the old extensions), this is fine as the right hand sidebar shows this now.
- At the bottom of each page the TOC (Table of contents) for all child packages or modules is displayed.  I think this is useful, especially if viewing on a smaller screen where the sidebar may be hidden.
- The order or sidebar contents are different but all are present
- The docstrings can be in a slightly different place but all are present I believe (I welcome someone doing a review to check some api doc pages)
- There is still a sphinxext (new one) that will remove the word **package** and **module** that make the sidebar hard to read imho.  I cannot find any option in **sphinx-apidoc** itself to disable this.
   - Without **package** and **module** in the sidebar (and headings)
      - image
  - With **package** and **module** in the sidebar (and headings)
     - **package** and **module** in the sidebar (and headings)
       - image

There are many incorrect indentations and incorrect syntax in the docstrings that should be using the [numpydocs](https://scitools-iris.readthedocs.io/en/latest/developers_guide/documenting/docstrings.html#docstrings), I may address those in a separate PR.

---
[Consult Iris pull request check list]( https://scitools-iris.readthedocs.io/en/latest/developers_guide/contributing_pull_request_checklist.html)

### CODE REVIEW

**Comment by trexfeathers** [comment]
OMG ❤

**Comment by rcomer** [comment]
😍

Think the linkcheck failure is https://github.com/matplotlib/matplotlib/issues/25731

**Comment by bjlittle** [comment]
> heart_eyes
> 
> Think the linkcheck failure is [matplotlib/matplotlib#25731](https://github.com/matplotlib/matplotlib/issues/25731)

There been a server outage for the `mpl` docs most of the afternoon, but it's back up now.

I respun the CI and all is well now :smile:

File: docs/src/whatsnew/latest.rst [hunk 1]
```diff
@@ -80,6 +80,10 @@ This document explains the changes made to Iris for this release
    ``livehtml`` to auto generate the documentation when changes are detected
    during development. (:pull:`5258`)
 
+#. `@tkknight`_ removed the legacy custom sphinx extensions that generates the
```
Comment (pp-mo) [inline docs/src/whatsnew/latest.rst:83]:
```suggestion
#. `@tkknight`_ removed the legacy custom sphinx extensions that generate the
```
tiny typo

**Review by pp-mo — APPROVED** [review]
Don't usually comment on docs PRs, but this one is a great improvement !
🚀

**Review by bjlittle — CHANGES_REQUESTED** [review]
@tkknight This is fantastic! 🥇 

I've been waiting for us to purge the custom extensions for many years... that said, they've served their purpose very well indeed.

Since #5274 landed, which also changed the lock files, could you rebase your PR and re-generate the lock files with your new `sphinxcontrib-apidoc` dependency 👍

File: requirements/locks/py39-linux-64.lock [hunk 2]
```diff
@@ -154,7 +154,7 @@ https://conda.anaconda.org/conda-forge/noarch/locket-1.0.0-pyhd8ed1ab_0.tar.bz2#
 https://conda.anaconda.org/conda-forge/linux-64/markupsafe-2.1.2-py39h72bdee0_0.conda#35514f5320206df9f4661c138c02e1c1
 https://conda.anaconda.org/conda-forge/linux-64/msgpack-python-1.0.5-py39h4b4f3f3_0.conda#413374bab5022a5199c5dd89aef75df5
 https://conda.anaconda.org/conda-forge/noarch/munkres-1.1.4-pyh9f0ad1d_0.tar.bz2#2ba8498c1018c1e9c61eb99b973dfe19
-https://conda.anaconda.org/conda-forge/linux-64/numpy-1.24.2-py39h7360e5f_0.conda#757070dc7cc33003254888808cd34f1e
+https://conda.anaconda.org/conda-forge/linux-64/numpy-1.23.5-py39h3d75532_0.conda#ea5d332e361eb72c2593cf79559bc0ec
```
Comment (pp-mo) [inline requirements/locks/py39-linux-64.lock:157]: note: same in all 3 lockfiles ?

**Pull request merged**